compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -668,8 +668,22 @@ impl<'p, 'tcx> MatchVisitor<'p, 'tcx> {
         let mut let_suggestion = None;
         let mut misc_suggestion = None;
         let mut interpreted_as_const = None;
+        let mut interpreted_as_const_sugg = None;
 
-        if let PatKind::Constant { .. }
+        if let PatKind::ExpandedConstant { def_id, is_inline: false, .. }
+        | PatKind::AscribeUserType {
+            subpattern:
+                box Pat { kind: PatKind::ExpandedConstant { def_id, is_inline: false, .. }, .. },
+            ..
+        } = pat.kind
+            && let DefKind::Const = self.tcx.def_kind(def_id)
+        {
+            let span = self.tcx.def_span(def_id);
+            let variable = self.tcx.item_name(def_id).to_string();
+            // When we encounter a constant as the binding name, point at the `const` definition.
+            interpreted_as_const = Some(span);
+            interpreted_as_const_sugg = Some(InterpretedAsConst { span: pat.span, variable });
+        } else if let PatKind::Constant { .. }
         | PatKind::AscribeUserType {
             subpattern: box Pat { kind: PatKind::Constant { .. }, .. },
             ..
```

Comment:
I think this `else` is redundant now: any named constant will give a `ExpandedConstant` so there's no need for this special-case that only applies to integers/strings and uses `span_to_snippet`.

Reply:
This branch is for [`let` bindings without an identifier (generally a literal)](https://github.com/rust-lang/rust/blob/81eef2d362a6f03db6f8928f82d94298d31eb81b/tests/ui/closures/2229_closure_analysis/bad-pattern.stderr#L14-L17) and we can't just remove it outright without losing output. We *can* remove the later inner `else` though (which might have been what you meant).